src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/Source.cs

```diff
@@ -0,0 +1,238 @@
+                participation: AsyncCompletionData.CompletionParticipation.ProvidesItems,
+                applicableToSpan: new SnapshotSpan(triggerLocation.Snapshot, service.GetDefaultCompletionListSpan(sourceText, triggerLocation.Position).ToSpan()));
+        }
+
+        public async Task<AsyncCompletionData.CompletionContext> GetCompletionContextAsync(
+            IAsyncCompletionSession session, 
+            AsyncCompletionData.CompletionTrigger trigger, 
+            SnapshotPoint triggerLocation, 
+            SnapshotSpan applicableToSpan, 
+            CancellationToken cancellationToken)
+        {
+            var document = triggerLocation.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return new AsyncCompletionData.CompletionContext(ImmutableArray<VSCompletionItem>.Empty);
+            }
+
+            var completionService = document.GetLanguageService<CompletionService>();
+
+            var completionList = await completionService.GetCompletionsAsync(
+                document,
+                triggerLocation,
+                GetRoslynTrigger(trigger, triggerLocation)).ConfigureAwait(false);
+
+            if (completionList == null)
+            {
+                return new AsyncCompletionData.CompletionContext(ImmutableArray<VSCompletionItem>.Empty);
+            }
+
+            var filterCache = new Dictionary<string, AsyncCompletionData.CompletionFilter>();
+
+            var itemsBuilder = new ArrayBuilder<VSCompletionItem>();
+            foreach (var roslynItem in completionList.Items)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                var item = Convert(document, roslynItem, completionService, filterCache);
+
+                // Have to store the snapshot to reuse it in some projections related scenarios
+                // where data and session in further calls are able to provide other snapshots.
+                item.Properties.AddProperty(TriggerSnapshot, triggerLocation.Snapshot);
+                itemsBuilder.Add(item);
+            }
+
+            var items = itemsBuilder.ToImmutableAndFree();
+
+            var suggestionItemOptions = completionList.SuggestionModeItem != null
+                    ? new AsyncCompletionData.SuggestionItemOptions(
+                        completionList.SuggestionModeItem.DisplayText,
+                        (completionList.SuggestionModeItem.Properties.TryGetValue(Description, out var description)
+                            ? description
+                            : string.Empty))
+                    : null;
+
+            // This is a code supporting legacy completion scenarios:
+            // If there are suggestionItemOptions, then later HandleNormalFiltering should set selection to SoftSelection.
+            session.Properties.AddProperty(HasSuggestionItemOptions, suggestionItemOptions != null);
```

Comment:
this seems weird.  won't we be able to tell this because the completioncontext contains the suggested item?